map_model/src/connectivity/walking.rs

```diff
@@ -125,7 +125,10 @@ pub fn all_walking_costs_from(
     // end of the sidewalk to the building
     for b in map.all_buildings() {
         if let Some(cost) = cost_per_node.get(&WalkingNode::closest(b.sidewalk_pos, map)) {
-            results.insert(b.id, *cost);
+            let distance_along_sidewalk= b.sidewalk_pos.dist_along();
+            let walking_duration_along_sidewalk = distance_along_sidewalk / opts.walking_speed;
```

Comment:
I don't understand this math; it assumes we always cross from the start of the sidewalk.

![Screenshot from 2021-01-03 09-54-07](https://user-images.githubusercontent.com/1664407/103485343-76edb100-4daa-11eb-9999-62e5c3fdba46.png)

Sidewalks are bidirectional, and the closest `WalkingNode` could be either end. In this example, the `dist_along` starts at 0 at the left intersection (and this is arbitrary; it just depends which direction the OpenStreetMap road is oriented originally). Say the red building is 30m from that start. The entire sidewalk is 50m. If `WalkingNode::closest` puts as the green X side, then the additional distance we want to travel is `50m - 30m`. If it put us at the other side, then it would just be the `30m`.

Inside `WalkingNode::closest`, we determine whether a positon is closer to the source intersection of a road (false) or destination (true) like this:
`let dst_i = map.get_l(pos.lane()).length() - pos.dist_along() <= pos.dist_along()`

So here, we need to do something like this:
```
let sidewalk_len = map.get_l(b.sidewalk()).length();
let bldg_dist = b.sidewalk_pos.dist_along();
let distance_from_closest_node = if sidewalk_len - bldg_dist <= bldg_dist {
  bldg_dist
} else {
  sidewalk_len - bldg_dist
};
let total_cost = *cost + distance_from_closest_node / opts.walking_speed;
```

Reply:
Thanks! I've implemented this approach in the latest revision.